network/p2p/tracer/gossipSubMeshTracer.go

```diff
@@ -187,13 +188,247 @@ func (t *GossipSubMeshTracer) Prune(p peer.ID, topic string) {
+	t.metrics.OnRpcReceived(msgCount, ihaveCount, iwantCount, graftCount, pruneCount)
+	if t.logger.GetLevel() == zerolog.TraceLevel {
+		t.logger.Trace().
+			Str("remote_peer_id", p2plogging.PeerId(p)).
+			Int("subscription_option_count", len(rpc.Subscriptions)).
+			Int("publish_message_count", msgCount).
+			Int("ihave_size", ihaveCount).
+			Int("iwant_size", iwantCount).
+			Int("graft_size", graftCount).
+			Int("prune_size", pruneCount).
+			Msg("sent pubsub rpc")
+	}
+
+	t.metrics.OnRpcSent(msgCount, ihaveCount, iwantCount, graftCount, pruneCount)
+}
+
+// AddPeer is called by GossipSub as a callback when a peer is added to the local node on a protocol, i.e., the local node is connected to the peer on a protocol.
+// The peer may or may not be subscribed to any topic.
+func (t *GossipSubMeshTracer) AddPeer(p peer.ID, proto protocol.ID) {
+	if t.logger.GetLevel() == zerolog.TraceLevel {
+		t.logger.Trace().
+			Str("local_peer_id", p2plogging.PeerId(p)).
+			Str("protocol", string(proto)).
+			Msg("peer added")
+	}
+	t.metrics.OnPeerAddedToProtocol(string(proto))
+}
+
+// RemovePeer is called by GossipSub as a callback when a peer is removed from the local node,
+// i.e., the local node is no longer connected to the peer.
+func (t *GossipSubMeshTracer) RemovePeer(p peer.ID) {
+	t.metrics.OnPeerRemovedFromProtocol()
+	if t.logger.GetLevel() == zerolog.TraceLevel {
+		t.logger.Trace().
+			Str("local_peer_id", p2plogging.PeerId(p)).
+			Msg("peer removed")
+	}
+}
+
+// Join is called by GossipSub as a callback when the local node joins a topic.
+func (t *GossipSubMeshTracer) Join(topic string) {
+	t.metrics.OnLocalPeerJoinedTopic()
+	if t.logger.GetLevel() == zerolog.TraceLevel {
+		t.logger.Trace().
+			Str("topic", topic).
+			Msg("local peer joined topic")
+	}
+}
+
+// Leave is called by GossipSub as a callback when the local node leaves a topic.
+func (t *GossipSubMeshTracer) Leave(topic string) {
+	t.metrics.OnLocalPeerLeftTopic()
+	if t.logger.GetLevel() == zerolog.TraceLevel {
+		t.logger.Trace().
+			Str("topic", topic).
+			Msg("local peer left topic")
+	}
+}
+
+// ValidateMessage is called by GossipSub as a callback when a message is received by the local node and entered the validation phase.
+// As the result of the validation, the message may be rejected or passed to the application (i.e., Flow protocol).
+func (t *GossipSubMeshTracer) ValidateMessage(msg *pubsub.Message) {
+	lg := t.logger.With().Logger()
+
+	if msg.Topic != nil {
+		lg = lg.With().Str("topic", *msg.Topic).Logger()
+	}
+	from, err := peer.IDFromBytes(msg.From)
+	if err == nil {
+		lg = lg.With().Str("remote_peer_id", p2plogging.PeerId(from)).Logger()
+	}
+	size := len(msg.Data)
+
+	t.metrics.OnMessageEnteredValidation(size)
+	if t.logger.GetLevel() == zerolog.TraceLevel {
+		lg.Trace().
+			Str("received_from", p2plogging.PeerId(msg.ReceivedFrom)).
+			Int("message_size", size).
+			Msg("received pubsub message entered validation phase")
+	}
+
```

Comment:
since `lg` is only used in the conditional, may as well abort early. same for the other 3 methods
```suggestion
	size := len(msg.Data)
	t.metrics.OnMessageEnteredValidation(size)
	if t.logger.GetLevel() > zerolog.TraceLevel {
		return
	}
	
	lg := t.logger.With().Logger()

	if msg.Topic != nil {
		lg = lg.With().Str("topic", *msg.Topic).Logger()
	}
	from, err := peer.IDFromBytes(msg.From)
	if err == nil {
		lg = lg.With().Str("remote_peer_id", p2plogging.PeerId(from)).Logger()
	}
	lg.Trace().
		Str("received_from", p2plogging.PeerId(msg.ReceivedFrom)).
		Int("message_size", size).
		Msg("received pubsub message entered validation phase")
```

Reply:
https://github.com/onflow/flow-go/pull/5131/commits/2321b384f74e299baf91b582eb5971ad989a720f